mlir/include/mlir/Dialect/Flux/Builder/FluxProgramBuilder.h

```diff
@@ -0,0 +1,404 @@
+   *
+   * @param qubit Input qubit (must be valid/unconsumed)
+   * @return Output qubit value
+   *
+   * @par Example:
+   * ```c++
+   * q = builder.reset(q);
+   * ```
+   * ```mlir
+   * %q_out = flux.reset %q : !flux.qubit -> !flux.qubit
+   * ```
+   */
+  Value reset(Value qubit);
+
+  //===--------------------------------------------------------------------===//
+  // Unitary Operations
+  //===--------------------------------------------------------------------===//
+
+  /**
+   * @brief Apply the X gate to a qubit
+   *
+   * @details
+   * Consumes the input qubit and produces a new output qubit SSA value.
+   * The input is validated and the tracking is updated.
+   *
+   * @param qubit Input qubit (must be valid/unconsumed)
+   * @return Output qubit
+   *
+   * @par Example:
+   * ```c++
+   * q_out = builder.x(q_in);
+   * ```
+   * ```mlir
+   * %q_out = flux.x %q_in : !flux.qubit -> !flux.qubit
+   * ```
+   */
+  Value x(Value qubit);
+
+  /**
+   * @brief Apply the RX gate to a qubit
+   *
+   * @details
+   * Consumes the input qubit and produces a new output qubit SSA value.
+   * The input is validated and the tracking is updated.
+   *
+   * @param theta Rotation angle
+   * @param qubit Input qubit (must be valid/unconsumed)
+   * @return Output qubit
+   *
+   * @par Example:
+   * ```c++
+   * q_out = builder.rx(1.0, q_in);
+   * ```
+   * ```mlir
+   * %q_out = flux.rx(1.0) %q_in : !flux.qubit -> !flux.qubit
+   * ```
+   */
+  Value rx(double theta, Value qubit);
+  Value rx(Value theta, Value qubit);
+
+  /**
+   * @brief Apply the U2 gate to a qubit
+   *
+   * @details
+   * Consumes the input qubit and produces a new output qubit SSA value.
+   * The input is validated and the tracking is updated.
+   *
+   * @param phi Rotation angle
+   * @param lambda Rotation angle
+   * @param qubit Input qubit (must be valid/unconsumed)
+   * @return Output qubit
+   *
+   * @par Example:
+   * ```c++
+   * q_out = builder.u2(1.0, 0.5, q_in);
+   * ```
+   * ```mlir
+   * %q_out = flux.u2(1.0, 0.5) %q_in : !flux.qubit -> !flux.qubit
+   * ```
+   */
+  Value u2(double phi, double lambda, Value qubit);
+  Value u2(double phi, Value lambda, Value qubit);
+  Value u2(Value phi, double lambda, Value qubit);
+  Value u2(Value phi, Value lambda, Value qubit);
+
+  /**
+   * @brief Apply the SWAP gate to two qubits
+   *
+   * @details
+   * Consumes the input qubits and produces new output qubit SSA values.
+   * The inputs are validated and the tracking is updated.
+   *
+   * @param qubit0 Input qubit (must be valid/unconsumed)
+   * @param qubit1 Input qubit (must be valid/unconsumed)
+   * @return Output qubits
+   *
+   * @par Example:
+   * ```c++
+   * {q0_out, q1_out} = builder.swap(q0_in, q1_in);
+   * ```
+   * ```mlir
+   * %q0_out, %q1_out = flux.swap %q0_in, %q1_in : !flux.qubit, !flux.qubit ->
+   * !flux.qubit, !flux.qubit
+   * ```
+   */
+  std::vector<Value> swap(Value qubit0, Value qubit1);
```

Comment:
I'd rather not use an std::vector here.
I believe the fitting type would be a std::pair, because that also allows you to use the syntax in the top level builder docstring with std::tie.